Variables docs (#7771)
garden-io/garden: docs/features/remote-variables.md

---
order: 20
title: Remote Variables and Secrets
---

## Overview

The Remote Variables feature allows you to store variables and secrets securely in [Garden Cloud](https://app.garden.io) and reference them in your Garden configuration. Remote variables and secrets can be scoped to environments and specific Garden users.

Here's a quick motivational example before we dive into the details. Below is a screenshot of secrets stored in Garden Cloud. Notice how the secrets are scoped to different environments and users 👇

<figure>
  <picture>
    <source
      srcset="https://public-assets-for-docs-site.s3.eu-central-1.amazonaws.com/remote-variables-dark.png"
      media="(prefers-color-scheme: dark)"
    />
    <img
      src="https://public-assets-for-docs-site.s3.eu-central-1.amazonaws.com/remote-variables-light.png"
      alt="A list of remote variables/secrets"
    />
  </picture>
  <figcaption>A list of remote variables/secrets</figcaption>
</figure>

> In your Garden config you can reference the `DB_PASSWORD` remote variable like so:

Reviewer: I'd be tempted to add a mention of the project config ahead of this, even though it's addressed below as well.

{% endhint %}

### Step 4 — Test that it works

First, make sure you're logged into Garden Cloud by running the login command from your Garden project:

```
garden login
```

Then verify that Garden can use the variables by running:

```console
garden get remote-variables
```

You can also do `garden get remote-variables -o json` for machine readable output.

You should see the variables just created in the output. If you created plain text variables, you'll see the value as well.

### Step 5 — Use them in your Garden config

You can now reference the variables you created anywhere in your Garden config with `${imported.<variable-name>}`. For example:

```
MY_VARIABLE: ${imported.MY_VARIABLE}
```

## Managing access with service accounts

{% hint style="danger" %}
Variables that aren't scoped to specific users are accessible to anyone in your organization. Read on to see how to manage access by scoping variables to user and/or service accounts.
{% endhint %}

Remote secrets can contain sensitive values that not everyone in your org should have access to. You can manage access by scoping them to specific users.

> A variable scoped to a user can not be used by other uses. Variables that are not scoped to users will be accessible to everyone in your Garden Cloud organization. Their values aren't visible if they're encrypted but users can still use them implicitly when running Garden commands.

Reviewer: ```suggestion
A variable scoped to a user can not be used by other user. Variables that are not scoped to users will be accessible to everyone in your Garden Cloud organization. Their values aren't visible if they're encrypted but users can still use them implicitly when running Garden commands.
```

The examples below assume you've read the Quickstart section above and that the `importVariables` config is set.

### Using remote variables in K8s manifests

We generally recommend using the `patchResources` field to override your K8s manifests as needed and this same pattern applies for remote variables. For example, this is how you'd set a remote variable as an environment variable:

```yaml
kind: Deploy
type: kubernetes
name: api
spec:
  manifestFiles: [my-manifests.yml]
  patchResources:
    - name: api # <--- The name of the resource to patch, should match the name in the K8s manifest
      kind: Deployment # <--- The kind of the resource to patch
      patch:
        spec:
          template:
            spec:
              containers:
                - name: api # <--- Should match the container name from the K8s manifest
                  env:
                    DB_PASSWORD: ${imported.DB_PASSWORD} # <--- You can define different values for different environments/users and Garden will resolve to the correct value.
```

For a more complete example of this approach, checkout our [K8s Deploy guide](../garden-for/kubernetes/deploy-k8s-resource.md#overwriting-values)

> ### Using remote variables in CI

Reviewer: This section feels like it could be folded in with the service account section above. Wouldn't you just use service accounts in CI anyway?

To use remote variables in CI you need to be authenticated against Garden Cloud via the `GARDEN_AUTH_TOKEN` environment variable.

We recommend creating a service account for CI environments—you'll find step-by-step instructions in the [Manage Access section](#managing-access-with-service-accounts) above. You can also create a personal auth token for your user from the Users page in [Garden Cloud](https://app.garden.io).

Once you have the auth token, you need to set in your CI environment as `GARDEN_AUTH_TOKEN`. If you're using GitHub Actions we recommend using the [Garden Action](https://github.com/garden-io/garden-action) which simplifies the process.

Storing variables in Garden Cloud, as opposed to only with your CI provider, allows you to fully reproduce a given CI run from your laptop. If your CI environment uses sensitive values that users should not have access to from their laptops, you can create a service account and scope the variables appropriately (see the [Managing Access section](#managing-access-with-service-accounts) above).

### Creating remote variables with the Garden CLI

You can programmatically create remote variables via the Garden CLI.

> First, get the variable list ID for the relevant list with. It's probably already visible in your project config under the `importVariables` field.

Reviewer: ```suggestion
First, get the variable list ID for the relevant list with. If you've already set the `importVariables` field in your project configuration, you can see the ID there.
```

You can also get all the variable lists with:

```sh
garden get variable-lists
```

Or `garden get variable-lists -o json` for a machine readable output.

Then create your remote variables with the `create remote-variables` command. For example:

```sh
garden create remote-variables varlist_123 DB_PASSWORD=my-pwd ACCESS_KEY=my-key

```

You can also create multiple variables at a time by passing a file of variables (dot env or JSON format) to the command. To see the different options, run:

```sh
garden create remote-variables --help
```

### Rotating variables that are about to expire

When creating variables you can optionally set an expiration date. For example if you create an access token in a platform you use with a three month lifetime you can include that information when creating the variable in Garden Cloud.

> You can then list all the variables that are about to expire with some `jq` magic (assuming you have `jq` installed):

Author: Ok this questionable and arguably the Garden command should allow you to do this "natively". But there's so many ways you might want to sort/filter that I figured I'd do show what's possible with jq.
